shark_turbine/kernel/wave/codegen.py

```diff
@@ -995,6 +996,79 @@ def handle_extract_slice(emitter: WaveEmitter, node: fx.Node):
     emitter.bind_node_proxy(node, IRProxyValue(element))
 
 
+###############################################################################
+# Reshape ops
+###############################################################################
+
+
+@handle_op(broadcast)
+def handle_broadcast(emitter: WaveEmitter, node: fx.Node):
+    try:
+        register, target_type = node.args
+    except ValueError as e:
+        raise ValidationError("Malformed arguments") from e
+    custom_arg = get_custom(register)
+    src_shape = get_custom(register).type.symbolic_shape
+    target_shape = target_type.symbolic_shape
+    hw_constraints = [
+        constraint
+        for constraint in emitter.constraints
+        if isinstance(constraint, HardwareConstraint)
+    ]
+    if len(hw_constraints) != 1:
+        raise NotImplementedError(
+            "Only support single HW Constraint for lowering broadcast."
+        )
+    hw_constraint = hw_constraints[0]
+    if not hasattr(hw_constraint, "vector_shapes"):
+        raise NotImplementedError(
+            "Only support broadcast for kernel with non-MMA constraints."
+        )
+
+    # As of writing of this pass, TKW expects only 1D vectors,
+    # or only 1 non-unit dim vector_shape. So we are not handling
+    # broadcast of N-d case.
+    broadcast_dims = set(target_shape) - set(src_shape)
```

Comment:
So this actually made me think about this broadcast op as a whole. I'm not sure you need to store [M, N] at all. Since we are forcing all vectors to be 1-d, you will always end up with vector<Qxf16> + vector<1xf16> (everything else will be disallowed). If that's the case, then we only need to keep track of what Q is in the broadcast op and I think you get that from elements_per_thread + propagation of those shapes. Let me know if I am missing something.

Reply:
That is a fantatstic comment/Q! In general, this is true but only implicitly, there is nothing in our language (from user perspective) that forces the vector_shapes to only have a single unit-dim, and there is no guarantees that we won't be supporting say 2D vectors in the future. 

TL;DR: Analysis is set to explicitly error out/safeguard unwanted broadcasts from happening, i.e if we do not check original shape say (M -> K,N) and if we implement 2D in the future, it will tell us what is missing/needs to be done to fix that.

Reply:
That makes sense and I think I see why you need to store [M, N]. Let me try and crystallize my thoughts.

Reply:
WRT just the original shape, we need the original shape accessible for expansion, register analysis, index_sequence_analysis as well

Reply:
So say we have a 3x4xi8 tensor with the following access patterns

1 1 2 2
3 3 4 4
5 5 6 6

And we want to add it to a 3xi8 tensor with the following access patterns

3
4
1

We can do it, but extract + splat won't give the right result.

On the other hand, if we had a 3xi8 with this access pattern

1
3
5

Now we could do an extract + splat which gives

1 1
3 3
5 5

This still doesn't give the right result. It only does if we the 3xi8 is
the result of a butterfly shuffle in which case threads 2, 4 and 6
have the same values as 1, 3 and 5 and so doing extract + splat ends up with

1 1 2 2
3 3 4 4
5 5 6 6

So this brings me to my first point which is we do have to maintain the tensor
dims so that we can compare the index sequences of the common dims to make
sure they are the same and either throw an error if not or emit shuffles.
We also have to keep note of whether the 1d tensor is the result of a butterfly
shuffle reduction and if not then throw an error or emit shuffles.

Now if we all the above conditions hold, then we need to do extract + splat.
For that, I think it would be easier if we just keep track of elements per
thread rather than recalculate it.

Reply:
I don't quite understand 3 4 1 access pattern very well, not sure if I understand the second case fully, why wouldn't the 2nd thread expand it's own data which => it doesn't need to shuffle data. Maybe we should chat over VC haha

Reply:
So 3 4 1 is just a sample random access pattern. What I was getting at is that in general the access patterns of the [N, M] tensor and [M] tensor might not be the same and so we need to add a check for this. Yes, sounds good :)

Reply:
Just double checking, if I understand your most recent comments correctly, we essentially need to do two things: 
1. we need to add more checks for the N,M case
2.refactor to use elements per thread instead of recomputing?

WRT 1, we can chat more over VC on Monday so we can draw out shapes and examples better. WRT 2, totally agree, although that means we'd need to start propagating elements_per_thread/ index sequence sizes. IIRC, currently we are only setting indexSequence size for reads/writes and MMA.